source/agora/node/Validator.d

```diff
@@ -0,0 +1,164 @@
+/*******************************************************************************
+
+    Implementation of the Validator API.
+
+    Copyright:
+        Copyright (c) 2019 - 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.Validator;
+
+import agora.api.Validator;
+import agora.common.Config;
+import agora.common.Hash;
+import agora.common.crypto.Key;
+import agora.common.Set;
+import agora.common.Task;
+import agora.common.Types;
+import agora.consensus.data.PreImageInfo;
+import agora.consensus.data.Transaction;
+import agora.consensus.protocol.Nominator;
+import agora.network.NetworkManager;
+import agora.node.FullNode;
+import agora.node.Ledger;
+import agora.utils.Log;
+import agora.utils.PrettyPrinter;
+
+import scpd.types.Stellar_SCP;
+
+import ocean.util.log.Logger;
+
+mixin AddLogger!();
+
+/*******************************************************************************
+
+    Implementation of the Validator node
+
+    This class implement the business code of the node.
+    Communication with the other nodes is handled by the `Network` class.
+
+*******************************************************************************/
+
+public class Validator : FullNode, API
+{
+    /// Nominator instance
+    protected Nominator nominator;
+
+    /// Ctor
+    public this (const Config config)
+    {
+        assert(config.node.is_validator);
+
+        // build the list of required quorum peers to connect to
+        Set!PublicKey required_peer_keys;
+        void getNodes (in QuorumConfig conf, ref Set!PublicKey nodes)
```

Comment:
I recommend moving this into a private static method called `buildRequiredKeys`. The constructor code is hard to follow when it has a big nested function like this one inside.

Reply:
Applied through additional commit.